### ISSUE TITLE
Interleaved groupnorm: legacy two-pass statistics corrupt sub-tile group widths

### ISSUE THREAD

**Issue by iwroszTT** [tenstorrent/tt-metal#53803]
### Summary

`ttnn.group_norm`'s legacy two-pass statistics path returns badly wrong values whenever groups are narrower than `TILE_WIDTH`, so that `input_mask` performs column selection inside a tile. The error grows monotonically with the number of groups and reaches **1.73 absolute on a normalized output whose true magnitude is 1.0** at an SD-UNet-like shape. Tile-aligned groups are unaffected.

This is a supported, documented configuration — `input_mask` exists precisely for it — and it fails silently. `use_welford=True` computes the same cases exactly, which localizes the defect to the legacy two-pass reduction rather than to addressing or masking.

Verified on `main` @ `f2a9207b30b` (2026-08-20), Blackhole p100a, `GroupNormNoMcastProgramFactory` with `device/kernels/compute/groupnorm.cpp`, grid `{1,1}`, bf16 TILE, `eps=1e-5`, golden = `torch.nn.functional.group_norm`.

CODEOWNERS: @tenstorrent/metalium-developers-mmfusedreduce

### Measurements

Input constructed so **every group has variance exactly 1** (each group alternates ±1 about a distinct mean), so `sqrt(var+eps) ≈ 1` and the numbers below are plain accuracy, not amplification of an ill-conditioned golden. Metric is max absolute error against the torch reference.

Sub-tile groups, scaling with group count:

| shape | num_groups | ch/group | legacy two-pass | `use_welford=True` |
|:---|---:|---:|---:|---:|
| `[1,1,32,32]` | 2 | 16 | 0.0391 | 0.0078 |
| `[1,1,32,64]` | 4 | 16 | 0.0781 | 0.0078 |
| `[1,1,32,128]` | 8 | 16 | **0.2734** | 0.0078 |
| `[1,1,32,256]` | 16 | 16 | **0.5566** | 0.0078 |
| `[1,1,1024,320]` | 32 | 10 | **1.7344** | guard-blocked, see below |

Tile-aligned groups (32 channels/group) are unaffected — flat at one bf16 ulp, `rows=32`:

| C | 32 | 64 | 96 | 128 | 256 |
|:---|---:|---:|---:|---:|---:|
| legacy two-pass | 0.0078 | 0.0078 | 0.0078 | 0.0078 | 0.0078 |

So the trigger is the sub-tile group width, not channel count, row count, or the factory.

### Localization

`use_welford=True` is exact on every case the legacy path fails, and flat at 0.0078 regardless of shape or group count. No compute-config knob rescues the legacy path. Measured on `[1,1,rows,64]`, `num_groups=2`, zero-variance groups (golden exactly 0), reported as `g0 / g1`:

| variant | 32 rows | 64 rows | 96 rows |
|:---|---:|---:|---:|
| default config | 0.0000 | 0.9492 / 0.9922 | 0.9766 / 0.9922 |
| `MathFidelity.LoFi` | 0.0000 | 0.9766 / 0.9766 | 1.0000 / 1.0000 |
| `MathFidelity.HiFi4` | 0.0000 | 0.9492 / 0.9922 | 0.9766 / 0.9922 |
| `fp32_dest_acc_en=True` | 0.0000 | 0.9844 / 0.9844 | 1.0000 / 1.0000 |
| HiFi4 + `fp32_dest_acc_en` | 0.0000 | 0.9766 / 0.0000 | 0.9609 / 0.9688 |
| **`use_welford=True`** | 0.0000 | **0.0000** | **0.0000** |

### Ruled out

- **Not the mask tensor.** Reproduced with `ttnn.create_group_norm_input_mask(C, num_groups, 1)` and, independently, with a hand-built mask verified element-identical to it.
- **Not the reader's column indexing.** For 16-channel groups, `num_channels_per_group % TILE_WIDTH = 16` is a power of two, so `reader_mcast_sender_unary_gn.cpp` takes the `GROUP_SIZE_IS_POWER_OF_2` branch at `:567-575`. With `row_offset` initialised to `num_cols_per_group = 16` and `block_w = 1`, the per-group `index_g_offset` sequence is 0, 0, 1, 1 — correct for four 16-channel groups over two channel tiles.
- **Not `num_out_blocks` splitting.** `groupnorm_heuristic_num_out_blocks` (`groupnorm_program_utils.cpp:56-69`) returns 1 for all the small shapes above.

### The Welford workaround is blocked exactly where the defect is worst

```
TT_FATAL @ groupnorm_no_mcast_program_factory.cpp:187: (!use_welford) || (num_groups_per_core <= 16)
num_groups_per_core (32) must be <= 16 when use_welfords is true. Increase the width of core grid to address this.
```

At `C=320, num_groups=32` on a `{1,1}` grid — the case measuring 1.73 error — Welford cannot be selected without widening the core grid. So "use Welford" is not a general mitigation.

### Merge Gate coverage

Nightly's interleaved groupnorm coverage is **ROW_MAJOR-input only**, so no existing test reaches interleaved TILE groupnorm at all. #53677 adds the first merge-gate coverage for these factories and carries two disabled repro cells in `tests/ttnn/unit_tests/gtests/test_normalization.cpp` (run with `--gtest_also_run_disabled_tests`):

- **`NormalizationSmoke.DISABLED_GroupNormInputMask`** — `[1,1,32,64]`, `num_groups=4`. This is the repro for **this** issue. Enable when fixed.
- **`NormalizationSmoke.DISABLED_GroupNormMultiRowTileWideC`** — `[1,1,64,64]`, `num_groups=2`. This is the repro for the sibling defect in the same code path, tracked separately (see Related below). Not a repro for this issue.

Note on the probe used by this issue's cell. It asserts an exact-0 golden on zero-variance groups, where the divisor is `sqrt(0+eps) = 0.00316` at `eps=1e-5`, so a mean off by one bf16 ulp is amplified about 300x and saturates to roughly ±1. That amplification is why the raw output magnitudes look identical (about 0.99) for defects of very different severity, and it is why the severity numbers in this issue are quoted on **variance-1** input instead. The cell is still a good repro — it is simply a sensitive detector, not a severity measure.

### Impact

- Silent wrong results in a supported configuration, degrading with group count — worst for the many-small-groups shapes real vision models use (`num_groups=32` is the diffusion-model default).
- **Blocks #43826.** That request asks `group_norm` to accept `C` that isn't a multiple of 32 (e.g. `C=120, num_groups=8` -> 15 channels/group). Unaligned `C` forces sub-tile group widths, i.e. exactly this path. Relaxing the guard without fixing this would convert a loud rejection into silently wrong numbers.

### Related but distinct

- #53846 — **inexact group mean in the same path**: for *tile-aligned* groups, the legacy two-pass reduction returns an inexact group mean at most tile counts. Probed with a single constant group (`C=32`, `num_groups=1`, `eps=1.0` so the divide is well conditioned), the mean error is exactly 0 at tiles-per-group ∈ {1, 4, 5, 8, 13, 16} and an integer multiple of 0.031006 — 1x, 2x, 3x, 4x or 5x — at the other 14 of the first 20 tile counts. Different trigger from this issue, which reproduces at **one** tile per group (a value that is exact there) and scales with group *count* instead. Welford is bit-exact for both. Likely a different fix (mask application vs the reduction), so tracked apart. Repro: `DISABLED_GroupNormMultiRowTileWideC`.
- #43826 — feature request for non-tile-aligned `C`; blocked by this.
- #51231 item 7 — `block_wt_last` declared `bool` in the sharded factory; different factory, different mechanism.

### Checklist

- [ ] Fix the legacy two-pass masked sub-tile-group statistics path
- [ ] Enable `GroupNormInputMask` in `test_normalization.cpp`
- [ ] Consider relaxing the `num_groups_per_core <= 16` Welford guard so Welford is a usable fallback
